chain/chain/src/store.rs

```diff
@@ -2028,6 +2028,43 @@ impl<'a> ChainStoreUpdate<'a> {
         Ok(())
     }
 
+    /// Clears chunk data which can be computed from other data in the storage.
+    ///
+    /// We are storing PartialEncodedChunk objects in the ColPartialChunks in
+    /// the storage.  However, those objects can be computed from data in
+    /// ColChunks and as such are redundant.  For performance reasons we want to
+    /// keep that data when operating at head of the chain but the data can be
+    /// safely removed from archival storage.
+    ///
+    /// `gc_stop_height` indicates height starting from which no data should be
+    /// garbage collected.  Roughly speaking this represents start of the ‘hot’
+    /// data that we want to keep.
+    ///
+    /// `gt_height_limit` indicates limit of how many heights to process.  This
+    /// limit means that the method may stop garbage collection before reaching
+    /// `gc_stop_height`.
+    pub fn clear_redundant_chunk_data(
+        &mut self,
+        gc_stop_height: BlockHeight,
+        gc_height_limit: BlockHeightDelta,
+    ) -> Result<(), Error> {
+        let start = self.chunk_tail()?;
+        let stop = gc_stop_height.min(start.saturating_add(gc_height_limit));
+        for height in start..stop {
+            let chunk_hashes = self.chain_store.get_all_chunk_hashes_by_height(height)?;
+            for chunk_hash in chunk_hashes {
+                let chunk_header_hash = chunk_hash.into();
+                self.gc_col(ColPartialChunks, &chunk_header_hash);
+                // Data in ColInvalidChunks isn’t technically redundant (as it
+                // cannot be calculated from any other data) but it is data we
+                // don’t need for anything so it can be deleted as well.
+                self.gc_col(ColInvalidChunks, &chunk_header_hash);
+            }
+        }
+        self.update_chunk_tail(stop);
```

Comment:
so here we actually start 'updating' chunk_tail for archival nodes, right?

are we 'abusing' it a little? shouldn't we use a separate 'pointer' for our garbage collection here? 

(as in theory we're removing only the partial chunks -- the real chunks are still there, so in theory the chunk_tail shouldn't have changed)

Reply:
I mean, I wouldn’t call it abuse, just repurposing. I’m not really sure if it’s worth adding a new pointer. I’ve talked with Bowen about turning of archival setting and we agreed that this is essentially unsupported operation so here we can just reuse the pointer. And in the future, e.g. with cold storage,n we’d need to make some custom migration step anyway at which point we’d know that data from genesis till chunk_tail is only partially garbage collected.

Reply:
> about turning off archival setting

this might be a nice starter project -- making sure that we 'fail' with a nice error if someone tries changing this config setting.

Reply:
https://github.com/near/nearcore/issues/6522